### PR TITLE
Added Testcases

### PR DESCRIPTION
**Pull request by TahiriSiamak**

### CODE REVIEW

File: src/Parser.php [hunk 1]
```diff
@@ -10,6 +10,11 @@ class Parser {
 	 * @throws	\InvalidArgumentException	if $input is not a valid exception
 	 * @return	Expression
 	 */
-	public function parse($input) {
+	public function parse(string $input) {
```
Comment (klees) [inline src/Parser.php]: This will only work with PHP 7.

File: tests/ParserTest.php [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+		$this->assertEquals($expression,$_expression);
+	}
+	public function test_Complex()
+	{
+		$parser = new Parser();
+		$expression = $parser->parse("2*3+4");
+		$_expression = new BinaryOp
+					( "*",
+					new Number(2),
+					new BinaryOp("+", new Number(3), new Number(4)));
```
Comment (klees) [inline tests/ParserTest.php]: I don't think, this is correct. 2*3+4 is meant to be (2*3)+4 by the rule "division/multiplication before subtraction/addition". You tree shows 2*(3+4)

File: tests/ParserTest.php [hunk 3]
```diff
@@ -0,0 +1,60 @@
+<?php
+
+class ParserTest extends PHPUnit_Framework_TestCase {
```
Comment (klees) [inline tests/ParserTest.php]: Please add a lot of tests here. There are simple tests missing. What should e.g. "3" when it is parsed? Tests for expression with brackets are missing too. You also need to tests some edge cases: What should e.g. 3-4+3 be when parsed? How about "3-4-5"? How should your parser deal with whitespaces? What should e.g. "3 + 4" be when parsed?

File: tests/ParserTest.php [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+		$this->assertEquals($expression,$_expression);
+	}
+	public function test_Complex()
+	{
+		$parser = new Parser();
+		$expression = $parser->parse("2*3+4");
+		$_expression = new BinaryOp
+					( "+",
+					new Number(4),
+					new BinaryOp("*", new Number(2), new Number(3)));
```
Comment (klees) [inline tests/ParserTest.php]: I think that these two lines need to be switched, because the multiplication comes on the left of the 4.

File: tests/ParserTest.php [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+		$this->assertEquals($expression,$_expression);
+	}
+	public function test_SingleExpression()
+	{
+		$parser = new Parser("3");
+		$new_number = new Number(3);
+		
+		$this->assertEquals(3, $new_number);
```
Comment (klees) [inline tests/ParserTest.php]: This will never work. Why?

File: tests/ParserTest.php [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+	}
+	public function test_Brackets()
+	{
+		$parser = new Parser();
+		$expression = $parser->parse("2*(3+4)");
+		$_expression = new BinaryOp
+					( "*",
+					new Number(2),
+					new BinaryOp("+", new Number(3), new Number(4)));
```
Comment (klees) [inline tests/ParserTest.php]: Nice.

File: tests/ParserTest.php [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+	}
+	public function test_EdgeCase1()
+	{
+		$parser = new Parser();
+		$expression = $parser->parse("3-4+3");
+		$_expression = new BinaryOp
+					( "-",
+					new Number(3),
+					new BinaryOp("+", new Number(4), new Number(3)));
+		$this->assertEquals(-4,$_expression);
```
Comment (klees) [inline tests/ParserTest.php]: Would you really interpret "3-4+3" as "3-(4+3)". I wouldn't. You could ask other people what the result of "3-4+3" is to find out what others think of this.

Comment (klees) [reply tests/ParserTest.php]: Also, the expression will never be equal to "-4".

File: tests/ParserTest.php [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+	}
+	public function test_EdgeCase2()
+	{
+		$parser = new Parser();
+		$expression = $parser->parse("3-4-5");
+		$_expression = new BinaryOp
+					( "-",
+					new Number(3),
+					new BinaryOp("-", new Number(4), new Number(5)));
```
Comment (klees) [inline tests/ParserTest.php]: Exactly like in the test case above, I doubt that people think "3-4-5" is "3-(4-5)".

File: tests/ParserTest.php [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+		$this->assertEquals(4,$_expression);
```
Comment (klees) [inline tests/ParserTest.php]: This will never be correct.

File: tests/ParserTest.php [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+	}
+	public function test_Spaces()
+	{
+		$parser = new Parser();
+		$expression = $parser->parse("3 + 4");
+		$expression = str_replace(" ", "", $expression);
```
Comment (klees) [inline tests/ParserTest.php]: What do you want to achieve with this line? It will completely whipe every expression the parser created before.

File: src/Parser.php [hunk 11]
```diff
@@ -10,6 +10,18 @@ class Parser {
 	 * @throws	\InvalidArgumentException	if $input is not a valid exception
 	 * @return	Expression
 	 */
-	public function parse($input) {
+	$input; 
+
+	public function Parser (string $input)
+	{
+		this->input = $input;
+	}
```
Comment (klees) [inline src/Parser.php]: What is this?

File: src/Parser.php [hunk 12]
```diff
@@ -10,6 +10,18 @@ class Parser {
...
+
+	public function parse(string $input) {
+		if($input=="")
+		{
+			throw new \InvalidArgumentException("Leere Eingabe");
+		}
+		//return Expression;
```
Comment (klees) [inline src/Parser.php]: Remove the body of this method for the moment.

File: tests/ParserTest.php [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+	}
+	public function test_EdgeCase1()
+	{
+		$parser = new Parser();
+		$expression = $parser->parse("3-4+3");
+		$_expression = new BinaryOp
+					( "+",
+					new Number(3),
+					new BinaryOp("-", new Number(3), new Number(4)));
+		$this->assertEquals(2,$_expression);
```
Comment (klees) [inline tests/ParserTest.php]: This will never be true.

File: tests/ParserTest.php [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+	}
+	public function test_EdgeCase2()
+	{
+		$parser = new Parser();
+		$expression = $parser->parse("3-4-5");
+		$_expression = new BinaryOp
+					( "-",
+					new Number(-5),
+					new BinaryOp("-", new Number(3), new Number(4)));
+		$this->assertEquals(-6,$_expression);
```
Comment (klees) [inline tests/ParserTest.php]: This will never be true.

File: tests/ParserTest.php [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+	}
+	public function test_Spaces()
+	{
+		$parser = new Parser();
+		$expression = $parser->parse("3 + 4");
+		$_expression = new BinaryOp
+					( " + ",
+					new Number(3),
+					new Number(4);
+		$this->assertEquals(7,$_expression);
```
Comment (klees) [inline tests/ParserTest.php]: This will never be true.

File: tests/ParserTest.php [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+	}
+	public function test_Whitespaces()
+	{
+		$parser = new Parser();
+		$expression = $parser->parse("3+/4");
+		$_expression = new BinaryOp
+					( "+/",
+					new Number(3),
+					new Number(4);
+		$this->assertEquals(7,$_expression);
```
Comment (klees) [inline tests/ParserTest.php]: This will never be true.

File: tests/ParserTest.php [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+	}
+	public function test_Whitespaces()
+	{
+		$parser = new Parser();
+		$expression = $parser->parse("3+/4");
+		$_expression = new BinaryOp
+					( "+/",
```
Comment (klees) [inline tests/ParserTest.php]: What kind of mathematical operator is this?

File: tests/ParserTest.php [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+	/*public function test_Whitespaces()
+	{
+		$parser = new Parser();
+		$expression = $parser->parse("3+/4");
+		$_expression = new BinaryOp
+					( "+/",
+					new Number(3),
+					new Number(4);
+		$this->assertEquals($expression,$_expression);
+	}*/
```
Comment (klees) [inline tests/ParserTest.php]: No uncommented code in the repos please. Either you need it or you don't, but you don't need to save it for later. This is what git does for us.

File: tests/ParserTest.php [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+					new Number(2)
+					);
+		$this->assertEquals($expression,$_expression);
+	}
+	public function test_Complex()
+	{
+		$parser = new Parser();
+		$expression = $parser->parse("2*3+4");
+		$_expression = new BinaryOp("*", new Number(2),
+					   new BinaryOp("+", new Number(3), new Number(4)));
```
Comment (klees) [inline tests/ParserTest.php]: This is not correct. This is (2\*3)+4 because multiplication before addition. Your tree is 2\*(3+4)

File: tests/ParserTest.php [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+		$this->assertEquals($expression,$_expression);
+	}
+	public function test_EdgeCase1()
+	{
+		$parser = new Parser();
+		$expression = $parser->parse("3-4+3");
+		$_expression = new BinaryOp
+					( "+",
+					new Number(3),
+					new BinaryOp("-", new Number(3), new Number(4)));
```
Comment (klees) [inline tests/ParserTest.php]: This is not correct. I still doubt that 3-4+3 is 3-(4+3), it rather is (3-4)+3. Your tree is none of these...

File: tests/ParserTest.php [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+		$this->assertEquals($expression, $_expression);
+	}
+	public function test_EdgeCase2()
+	{
+		$parser = new Parser();
+		$expression = $parser->parse("3-4-5");
+		$_expression = new BinaryOp
+					( "-",
+					new Number(-5),
+					new BinaryOp("-", new Number(3), new Number(4)));
```
Comment (klees) [inline tests/ParserTest.php]: I fail to these how this tree represents the given string.

File: tests/ParserTest.php [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+					new Number(2)
+					);
+		$this->assertEquals($_expression,$expression);
+	}
+	public function test_Complex()
+	{
+		$parser = new Parser();
+		$expression = $parser->parse("2*3+4");
+		$_expression = new BinaryOp("+", new Number(4),
+					   new BinaryOp("*", new Number(2), new Number(3)));
```
Comment (klees) [inline tests/ParserTest.php]: 2\*3 is on the left of the + in the expression but on the right in the tree you build.

File: tests/ParserTest.php [hunk 23]
```diff
@@ -0,0 +1,118 @@
...
+		$this->assertEquals($_expression,$expression);
+	}
+	public function test_EdgeCase1()
+	{
+		$parser = new Parser();
+		$expression = $parser->parse("3-4+3");
+		$_expression = new BinaryOp
+					( "+",
+					new Number(3),
+					new BinaryOp("-", new Number(3), new Number(4)));
```
Comment (klees) [inline tests/ParserTest.php]: 3-4 is on the left of the + in the expression, but on the right in the tree you build.

File: tests/ParserTest.php [hunk 24]
```diff
@@ -0,0 +1,118 @@
...
+		$this->assertEquals($_expression, $expression);
+	}
+	public function test_EdgeCase2()
+	{
+		$parser = new Parser();
+		$expression = $parser->parse("3-4-5");
+		$_expression = new BinaryOp
+					( "-", 
+					new Number(5),
+					new BinaryOp("-", new Number(3), new Number(4)));
```
Comment (klees) [inline tests/ParserTest.php]: 5 is on the left of the - in the expression but on the right of the minus in the tree you build.
